### PR TITLE
feat(linter): react/exhaustive-deps

### PR DESCRIPTION
**Pull request by camc314**
Firstly, a massive thanks to @alisnic for starting this (incredibly complicated) lint rule in https://github.com/oxc-project/oxc/pull/2637 !

closes  #2637
relates to #2174

### CODE REVIEW

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camc314) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:644]: any ideas on how to get around this?

Comment (camc314) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:672]: @overlookmotel / @Boshen  ? 🙂

**Comment by camc314** [comment]
Rough first draft but most of the way their. would appreciate some early reviews as this was a massive amount of work and i don't want to go too far 🙂

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (Boshen) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:709]: Store the name / span / id directly inside `Dependency`?

**Comment by Boshen** [comment]
> Rough first draft but most of the way their. would appreciate some early reviews as this was a massive amount of work and i don't want to go too far 🙂

This is amazing! I would run it through oxlint ecosystem ci and see what happens.

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:48]: Should this be help text?
```suggestion
    OxcDiagnostic::warn(format!(
      "React Hook {hook_name} received a function whose dependencies are unknown."
    ))
    .with_help("Pass an inline function instead.")
    .with_label(span)
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:53]:
```suggestion
    OxcDiagnostic::warn("Effect callbacks are synchronous to prevent race conditions.")
        .with_help("Remove `async` from the callback to make it synchronous.")
        .with_label(span)
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:96]:
```suggestion
    OxcDiagnostic::warn("The literal is not a valid dependency because it never changes.")
        .with_help("Remove the literal from the array.")
        .with_label(span)
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:103]:
```suggestion
    OxcDiagnostic::warn(format!(
        "React Hook {hook_name} has a complex expression in the dependency array.",
    ))
    .with_help("Extract the expression to a separate variable so it can be statically checked.")
    .with_label(span)
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:110]: We should suggest memoizing I think, bonus points if we can point to the original definition of the variable with a span
```suggestion
    OxcDiagnostic::warn(format!(
        "React Hook {hook_name} has a dependency array that changes every render."
    ))
    .with_help("Try memoizing this variable with `useRef` or `useCallback`.")
    .with_label(span)
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:122]:
```suggestion
    .with_help("Consider removing it from the dependency array. Outer scope values like aren't valid dependencies because mutating them doesn't re-render the component.")
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:645]:
```suggestion
            if let Expression::Identifier(_) = &member.object {
```

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camc314) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:709]: good shout 🙂

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:681]: I would recommend avoiding creating a vec here, can we push to a string directly instead using `String.push` and `String.push_str`?

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camchenry) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:1102]: there are enough tests here that it might be a good idea to move it into a separate file? not sure what our standards here though are @Boshen

**Review by camchenry — COMMENTED** [review]
This looks pretty good overall, I didn't thoroughly review all of the code (there is a lot!) but I took a look at the diagnostics and have a few help text suggestions. I think keeping this in nursery and continuing to iterate on it would be good

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camc314) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:54]: the react help is a lot more complex here
i changed the help to `"Consider putting the asynchronous code inside a function and calling it from the effect."`

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camc314) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:686]: updated to use an iterator
```rs
std::iter::once(&self.iref.name)
    .chain(self.chain.iter())
    .map(|s| s.as_str())
    .join(".")
```

which should avoid the vec?

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (Boshen) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:1110]: 1 file is fine, makes it easy to search for things.

**Comment by Boshen** [comment]
Note: once merged, manually add this rule to oxlint-ecosystem-ci's CI script. We then change this to correctness after running it for 2 - 4 weeks.

File: crates/oxc_linter/src/service/runtime.rs
Comment (camchenry) [inline crates/oxc_linter/src/service/runtime.rs:210]: I gave it a shot, I think this works:

```rust
let should_build_child_ids = self.linter.rules.iter().any(|rule| matches!(rule.rule, RuleEnum::ExhaustiveDeps(_)));
```

```suggestion
            .with_scope_tree_child_ids(should_build_child_ids)
```

**Review by camchenry — APPROVED** [review]
it's hard to review all of this code, but I have skimmed over most of it and not too much stuck out to me. I think this is ready to start being tested in nursery

File: crates/oxc_linter/src/service/runtime.rs
Comment (camc314) [reply crates/oxc_linter/src/service/runtime.rs:210]: hmm let me get the benchmark working and look at the perf hit (if any)

**Comment by camc314** [comment]
@Boshen did it look ok in the ecosystem ci? https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/11755920122/job/32751524336

File: crates/oxc_linter/src/service/runtime.rs
Comment (camc314) [reply crates/oxc_linter/src/service/runtime.rs:210]: Screenshot 2024-11-09 at 15 58 35

shall we leave for now? look like there's no perf change


(i'm just about to put a comit to disable running this rule on non-`tsx/jsx` files to fix the checker.ts perf regression)

Comment (camchenry) [reply crates/oxc_linter/src/service/runtime.rs:210]: It might just be random, that doesn't seem like a big perf change (if any) to me so I think it's fine

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camc314) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:550]: can anyone think of a case where we would want this to run on non-jsx?

Comment (camc314) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:550]: hmm i guess this is incorrect in the case where people are using `createElement` rather than jsx

Comment (camchenry) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs]: Also lib files can be .ts files that have hooks in them with effects like https://github.com/streamich/react-use/

Comment (camc314) [reply crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs]: ah yes good thought

**Comment by Boshen** [comment]
### Merge activity

* **Nov 11, 3:36 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

**Comment by Boshen** [comment]
Hit the merge button too quickly 😅 

In `oxlint-ecosystem-ci/repos`:

```
~/oxc/target/release/oxlint -A all -W exhaustive-deps
Finished in 5.5s on 117561 files with 1 rules using 8 threads.
Found 536 warnings and 23 errors.
```

Some false positives:

```
affine  canary ❯ ~/oxc/oxc/target/release/oxlint -A all -W exhaustive-deps

  ⚠ react_hooks(exhaustive-deps): React Hook useMemo has a missing dependency: 'T'
     ╭─[packages/frontend/core/src/components/page-list/virtualized-list.tsx:142:6]
 141 │     return items;
 142 │   }, [groupCollapsedState, groups]);
     ·      ─────────────────────────────
 143 │ };
     ╰────
  help: Either include it or remove the dependency array.
```

It's picking up time annotation

```
    const items: VirtuosoItem<T>[] = [];
```

**Comment by Boshen** [comment]
Do you also need to include function calls?

```
  ⚠ react_hooks(exhaustive-deps): React Hook useEffect has a missing dependency: 'makeMappingRequest'
    ╭─[kibana/x-pack/plugins/enterprise_search/public/applications/enterprise_search_content/components/search_index/index_error.tsx:83:6]
 82 │     makeMappingRequest({ indexName });
 83 │   }, [indexName]);
    ·      ───────────
 84 │
    ╰────
  help: Either include it or remove the dependency array.
```

**Comment by Boshen** [comment]
```
  ⚠ react_hooks(exhaustive-deps): React Hook useCallback has unnecessary dependency: dataVisualizerListState.sortField
     ╭─[kibana/x-pack/plugins/data_visualizer/public/application/index_data_visualizer/hooks/use_field_stats.ts:307:5]
 306 │     dataVisualizerListState.sortDirection,
 307 │     dataVisualizerListState.sortField,
     ·     ───────────────────────
 308 │     samplingProbability,
     ╰────
  help: Either include it or remove the dependency array.
```

The span is not pointing at the member expression.

**Comment by Boshen** [comment]
This looks pretty weird

```
  ⚠ react_hooks(exhaustive-deps): React Hook useEffect has missing dependencies: 'searchParams.get', 't', and 'form.setValue'
    ╭─[cal/packages/features/eventtypes/components/DuplicateDialog.tsx:67:6]
 66 │     }
 67 │   }, [searchParams?.get("dialog")]);
    ·      ─────────────────────────────
 68 │
    ╰────
  help: Either include it or remove the dependency array.
```

**Comment by Boshen** [comment]
kibana has the most violations, you may want to test it during developement.

**Comment by camc314** [comment]
> This looks pretty weird
> 
> ```
>   ⚠ react_hooks(exhaustive-deps): React Hook useEffect has missing dependencies: 'searchParams.get', 't', and 'form.setValue'
>     ╭─[cal/packages/features/eventtypes/components/DuplicateDialog.tsx:67:6]
>  66 │     }
>  67 │   }, [searchParams?.get("dialog")]);
>     ·      ─────────────────────────────
>  68 │
>     ╰────
>   help: Either include it or remove the dependency array.
> ``

🙂 looks like we're following the same behaviour as eslint


Screenshot 2024-11-11 at 09 40 29

File: crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs
Comment (camc314) [inline crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs:882]: TODO: fix

https://github.com/oxc-project/oxc/pull/7151#issuecomment-2467554800

https://github.com/facebook/react/blob/1b0132c05acabae5aebd32c2cadddfb16bda70bc/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L265-L320

**Comment by camc314** [comment]
> s/search_index/index_error.tsx:83:6

looks like they've got it disabled? for this file https://github.com/elastic/kibana//blob/2466a172af66452bdd939dddc56506faba7ffb7a/.eslintrc.js#L1623

when i run eslint on this file, it reports (same as us)

File: crates/oxc_linter/src/snapshots/exhaustive_deps.snap
Comment (camc314) [reply crates/oxc_linter/src/snapshots/exhaustive_deps.snap]: fixed https://github.com/oxc-project/oxc/pull/7151/commits/e9144c31b1061d6c35861c2c02c37f919b057fc8

**Comment by camc314** [comment]
@Boshen after https://github.com/oxc-project/oxc/pull/7151/commits/a4f7675ef509ade89ccbc524bf5f2a80bb5f187c i don't there are more false positives in affine at least

**Comment by Boshen** [comment]
```
  ⚠ react_hooks(exhaustive-deps): React Hook useCallback received a function whose dependencies are unknown.
    ╭─[kibana/src/plugins/console/public/application/hooks/use_save_current_text_object.ts:27:5]
 26 │       return useCallback(
 27 │ ╭─▶     throttle(
 28 │ │         (text: string) => {
 29 │ │           const { current: promise } = promiseChainRef;
 30 │ │           if (!currentTextObject) return;
 31 │ │           promise.finally(() =>
 32 │ │             objectStorageClient.text.update({ ...currentTextObject, text, updatedAt: Date.now() })
 33 │ │           );
 34 │ │         },
 35 │ │         WAIT_MS,
 36 │ │         { trailing: true }
 37 │ ╰─▶     ),
 38 │         [objectStorageClient, currentTextObject]
    ╰────
  help: Pass an inline function instead.
 ```
 
Interesting case, how should this be handled?

**Review by Boshen — APPROVED** [review]
Feel free to merge after a final self review.

**Comment by camc314** [comment]
> /use_save_current_text_object.ts:27:5



> ```
>   ⚠ react_hooks(exhaustive-deps): React Hook useCallback received a function whose dependencies are unknown.
>     ╭─[kibana/src/plugins/console/public/application/hooks/use_save_current_text_object.ts:27:5]
>  26 │       return useCallback(
>  27 │ ╭─▶     throttle(
>  28 │ │         (text: string) => {
>  29 │ │           const { current: promise } = promiseChainRef;
>  30 │ │           if (!currentTextObject) return;
>  31 │ │           promise.finally(() =>
>  32 │ │             objectStorageClient.text.update({ ...currentTextObject, text, updatedAt: Date.now() })
>  33 │ │           );
>  34 │ │         },
>  35 │ │         WAIT_MS,
>  36 │ │         { trailing: true }
>  37 │ ╰─▶     ),
>  38 │         [objectStorageClient, currentTextObject]
>     ╰────
>   help: Pass an inline function instead.
> ```
> 
> Interesting case, how should this be handled?

lol we do the same thing as react, only we're currently reporting on a different line so it wasn't being ignored


Screenshot 2024-11-12 at 08 54 34

**Comment by camc314** [comment]
@Boshen i guess we should be reporting the same (or at least very similar spans to react? if ok with you, i'll make an issue to track bits that aren't exactly complete for this rule

**Comment by Boshen** [comment]
> @Boshen i guess we should be reporting the same (or at least very similar spans to react? if ok with you, i'll make an issue to track bits that aren't exactly complete for this rule

Indeed, current span breaks already written eslint ignores.

**Pull request merged**
